### PR TITLE
feat(core): make event track great again

### PR DESCRIPTION
**Pull request by forehalo**

### CODE REVIEW

**Comment by forehalo** [comment]
* **#7668** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7668?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7667** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7667?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7664** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7664?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7661** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7661?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7735** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7735?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7695** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7695?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @forehalo and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/core/src/mixpanel/events/type.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      [Module in keyof MixpanelEvents[Page][Segment]]: {
+        [Control in keyof MixpanelEvents[Page][Segment][Module]]: {
+          [Event in keyof MixpanelEvents[Page][Segment][Module][Control]]: (
+            argv: MixpanelEvents[Page][Segment][Module][Control][Event]
+          ) => void;
+        };
+      };
+    };
+  };
+};
```
Comment (EYHN) [inline packages/frontend/core/src/mixpanel/events/type.ts]:
```suggestion
type KeysOfUnion<T> = T extends T ? keyof T : never;
type ValuesOfUnion<T> = T extends T ? T[keyof T] : never;

type MakeEventLevel<
  Level extends string = '>',
  Type = MixpanelEvents,
> = Level extends '>>>>>>'
  ? {
      [Key in KeysOfUnion<Type>]: (
        argv: Extract<Type, Record<Key, any>>[Key]
      ) => void;
    }
  : {
      [Key in KeysOfUnion<Type> | '$']: MakeEventLevel<
        `${Level}>`,
        Key extends '$'
          ? ValuesOfUnion<Type>
          : Extract<Type, Record<Key, any>>[Key]
      >;
    };

type MakeEvent = MakeEventLevel;
```

Comment (EYHN) [reply packages/frontend/core/src/mixpanel/events/type.ts]: We can then replace any unknown part with $
```ts
track.$.cmdk.$.quickSearchOption.changeLanguage()
```

Comment (forehalo) [reply packages/frontend/core/src/mixpanel/events/type.ts]: the typing is gonna be in a mess

Comment (EYHN) [reply packages/frontend/core/src/mixpanel/events/type.ts]: Works fine on my side

File: packages/frontend/core/src/components/root-app-sidebar/index.tsx [hunk 2]
```diff
@@ -158,6 +154,7 @@ export const RootAppSidebar = (): ReactElement => {
           <QuickSearchInput
             className={quickSearch}
             data-testid="slider-bar-quick-search-button"
+            data-event-props="$.navigationPanel.generalFunction.quickSearch"
```
Comment (EYHN) [inline packages/frontend/core/src/components/root-app-sidebar/index.tsx:156]: The original definition is to track when opening. It seems that every mouse click will track here?

Comment (forehalo) [reply packages/frontend/core/src/components/root-app-sidebar/index.tsx:156]: what do you mean by every click?

Comment (EYHN) [reply packages/frontend/core/src/components/root-app-sidebar/index.tsx:156]: sorry, I misunderstood

File: packages/frontend/core/src/mixpanel/events.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
+// let '$' stands for unspecific matrix
+/* eslint-disable rxjs/finnish */
+export interface Events {
```
Comment (pengx17) [inline packages/frontend/core/src/mixpanel/events.ts:3]: IMO it seems it is a bit hard to figure out the relationship between layers to module/page/segment etc.
Not sure if it is better/possible to divide this monolith definition into pieces?

**Review by pengx17 — COMMENTED** [review]
Great abstraction! Can you please write a RFC for it?

File: packages/frontend/core/src/mixpanel/events.ts (hunk 3 above)
Comment (forehalo) [reply packages/frontend/core/src/mixpanel/events.ts:3]: I tried but it only went more verbose. I will make it clear regarding the layers after tests

**Review by pengx17 — APPROVED** [review]

**Pull request merged**